### PR TITLE
Set up type checking using Sorbet and annotate parts of the code.

### PR DESCRIPTION
**Pull request by mattmccutchen**
This PR sets up the infrastructure for static type checking for the Braid codebase and annotates several parts of the code as an example.  I can annotate more of the code later, and we can discuss whether refactoring the code to make more of it statically typable is justified.

@realityforge I realize this is an ambitious proposal to spring on you, but I hope that if you take your time to review it, you'll come to agree that it's the right thing to do for the long-term maintainability of Braid.  (Maybe I should have asked you in an issue first, but I thought the PR would be easy enough to just submit, until I ran into more and more problems. :slightly_frowning_face: )  I've tried to make this change as low-risk as possible, as further described in the commit message and included documentation.  I hope you'll review this with your greater knowledge of how things are customarily done in Ruby and make any suggestions for how things could be set up better, with the caveat that I'd like to keep this PR focused on enabling type checking and defer other refactorings to separate PRs whenever reasonable.

Ever since I started working on Braid, I've struggled to navigate and understand the code so I could modify it without breaking things.  Sometimes I still made silly mistakes that broke things; you'll see one such bug that I found while working on this PR.  Static type information and type checking can help a lot with this by enabling navigation to definitions and references and catching a significant fraction of silly mistakes in real time via red underlines in the IDE.  This is even more valuable for contributors with a weak understanding of Ruby's semantics, since many mistakes due to misunderstanding of Ruby semantics will be caught in real time.  I've gotten more familiar with Ruby over time from working on Braid, but there are still some advanced Ruby features that I'm liable to mess up.  I'm hoping to do some more Braid feature development soon, and I would strongly prefer to have type checking set up first.

[Sorbet](https://sorbet.org/) seems to be the only option for Ruby type checking right now, so that's what I'm using.  It's not very mature and I can't be sure it will continue to be developed long-term, but I think it brings enough value even in its current form to justify the time investment of writing Sorbet annotations.

The test suite passed on my Linux and Windows machines.  I trust you'll run the test suite on Mac OS.  I'd encourage you to also try out the IDE support on whichever OS supported by the Sorbet static analyzer (Linux or Mac OS) you prefer for development and also address the TODO comment I left for you in the `Gemfile` to add the proper logic to enable the static analyzer on Mac OS, even if you don't plan to use it yourself. :slightly_smiling_face:

### CODE REVIEW

**Comment by realityforge** [comment]
Interesting ... I have never encountered Sorbet before. It looks interesting. I may try to get to looking at it but it may be a week or so. 

About the only thing that jumped out of me was the work around for method_missing idiom. We should just remove the magic and explicitly manage the state.

Poking about the Jetbrains IDE suite has slightly more advanced support for sorbet so I will test it when I get to it. https://blog.jetbrains.com/ruby/2021/10/rubymine-2021-3-eap2-initial-support-for-sorbet-improved-debugger-and-more/

File: lib/braid/mirror.rb [hunk 1]
```diff
@@ -176,56 +210,79 @@ def fetch_base_revision_if_missing
...
+    sig {returns(T::Boolean)}
     def cached?
       git.remote_url(remote) == cached_url
     end
 
+    sig {returns(String)}
     def base_revision
-      if revision
+      # Avoid a Sorbet "unreachable code" error.
+      # TODO (typing): Is the revision expected to be non-nil nowadays?  Can we
```
Comment (realityforge) [inline lib/braid/mirror.rb:227]: Yes revision and url are required to be non-null. However, eyeballing the code seems to suggest we don't enforce this ... which I am curious about. However it is marked as required in schema. Hmmm.

We should be able to delete the inferred_revision branch (and the inferred_revision method) but it is difficult to verify without digging into the code more than I have time to do right now

**Review by realityforge — APPROVED** [review]
It seems like it is a positive move, even if I am not familiar with it. It works on OSX and within the few IDEs I have tried (RubyMine and IntelliJ IDEA with ruby plugin). So it seems to be providing value.

To be honest I have often thought about rewriting braid in a statically typed language (i.e. rust or haskell) as a way of learning a new language but this seems like a much easier path thats is more likely to occur ;-)

**Review by mattmccutchen — COMMENTED** [review]
After ~1.5 months, I think I finally have your comments satisfactorily addressed, and I fixed some other style issues I noticed as well.  I have limited time to work on Braid, but I can get things done eventually. :slightly_smiling_face:

I merged `master` into this PR branch to more easily test the PR in combination with the other recent changes.  So if this PR is merged with a "merge commit", that will leave an additional merge commit in the history that may make the history marginally more difficult for people to understand later.  How do you feel about this?  One alternative is to "squash merge", which leaves a simpler history on `master` but means anyone interested has to go to GitHub to view the fine-grained history of the PR, and if the repository is ever deleted from GitHub, that history may be lost unless we take extra steps to archive it (though that would probably be less of a problem than the loss of other data that's currently only on GitHub, such as the open issues).  Or I can regenerate the PR branch without the merge, and we can construct a temporary merge to run tests before merging the PR; I think some projects do that, though it's trickier IMO.

> About the only thing that jumped out of me was the work around for method_missing idiom. We should just remove the magic and explicitly manage the state.

Done.  That turned out to be a good idea: it gave us runtime checks for the accessors in the test suite, which caught an incorrect annotation.

**Comment by realityforge** [comment]
Sorry I haven't got to look at this for a bit. Life is a bit hectic atm and it looks like I will not get to this until next week. Sorry again!

File: Gemfile [hunk 2]
```diff
@@ -1,3 +1,76 @@
...
+group :development do
+  gem 'sorbet-runtime', SORBET_VERSION_SPEC
+
+  # Note: `install_if` is supported only by Bundler and not by `gem install -g`,
+  # which users may prefer due to its `--conservative` option (for which Bundler
+  # doesn't seem to have an analogue), so we use a plain Ruby `if` instead.
+  #
+  # TODO for realityforge: Please add the correct logic for your macOS test
+  # system here. :)
+  if RUBY_PLATFORM == 'x86_64-linux' then
```
[the remark this answers is not in the document]
Comment (realityforge) [reply Gemfile]: Done ;-)

File: lib/braid/operations.rb [hunk 3]
```diff
@@ -166,6 +166,7 @@ def relative_working_dir
 
       def commit(message, *args)
         cmd = 'git commit --no-verify'
+        message_file = nil
```
[the remark this answers is not in the document]
Comment (realityforge) [reply lib/braid/operations.rb:169]: Both RubyMine and IntelliJ with the Ruby plugin will detect these issues. While you can run it from the command line ... and I have done so my installing it into a docker image. I believe the work to do so is probably not worth the effort ;-)

There are two "errors" that it detected in the IDE with the current typing. These are both in `mirror.rb`

Screen Shot 2022-06-09 at 1 30 21 pm
Screen Shot 2022-06-09 at 1 30 43 pm

**Review by realityforge — APPROVED** [review]

**Comment by realityforge** [comment]
I think this is "good enough" to merge to master. It is better than what we currently have ;-) I have left the PR open so you can review my changes in it but when you are satisfied then merge away.

**Review by mattmccutchen — COMMENTED** [review]
I want to make one more round of changes to address some inconsistency in your latest changes.  Hopefully this will be the last.  The tests pass on my Linux and Windows systems.

One thing we still need to decide is the merge method for the PR, per [my previous comment](https://github.com/cristibalan/braid/pull/105#pullrequestreview-951226581).  If we do a squash merge, I envision something like the following for the squash commit message:
```
Set up type checking + be stricter about invocation + other cleanup.

- Set up type checking using Sorbet and annotate parts of the code.  See
  the new "Type checking" section of development.md for a full
  explanation.  To make this change lower-risk for end users, Braid
  doesn't depend on Sorbet at all during normal use by end users; Sorbet
  runtime type checks are performed only by the test suite.

- Require Braid code to be run with the correct gems already activated.
  This ensures that the correct versions of Braid and its dependencies
  are used, which is important now that we have a dependency (Sorbet)
  that reserves the right to make breaking changes in any update.  This
  change should make no difference to end users using an installed Braid
  gem.  However, Braid developers will now have to go through Bundler
  when running the `braid` executable or commands such as rake or rspec.

- Various other cleanups.

See https://github.com/cristibalan/braid/pull/105 for more details and
rationale.

Co-authored-by: Peter Donald <peter@realityforge.org>
```
However, I lean toward making a merge commit and keeping the full history.  While it's a bit messy, I think the Braid project is still at a small enough scale that readers should be able to handle it, and the finer breakdown and additional information in the commit messages could be useful to some readers.

File: lib/braid/config.rb [hunk 4]
```diff
@@ -1,6 +1,8 @@
+# typed: true
 require 'yaml'
 require 'json'
 require 'yaml/store'
+require 'braid/sorbet/setup'
```
[the remark this answers is not in the document]
Comment (realityforge) [reply lib/braid/config.rb]: > However, I lean toward making a merge commit and keeping the full history. While it's a bit messy, I think the Braid project is still at a small enough scale that readers should be able to handle it, and the finer breakdown and additional information in the commit messages could be useful to some readers.

Seems reasonable to me.



> Why do this here? In general, we seem to rely on `lib/braid.rb` to `require`

An accident ;-) I think I attempted to jump straight to strict type checking and this allowed me to get further but after fixing other bugs I realised this was not required but never got back to removing it.

> I'm tentatively removing this line.

👍

File: lib/braid/config.rb [hunk 5]
```diff
@@ -214,11 +233,12 @@ def write_db
 
     private
 
+    sig {params(config_file: String, old_config_files: T::Array[String]).returns(TODO_TYPE)}
     def load_config(config_file, old_config_files)
       (old_config_files + [config_file]).each do |file|
         next unless File.exist?(file)
         begin
-          store = YAML::Store.new(file)
+          store = T.let(YAML::Store, T.untyped).new(file)
```
[the remark this answers is not in the document]
Comment (realityforge) [reply lib/braid/config.rb:240]: Ok. I just missed the intent of `TODO_TYPE`. However I am happy to go with whatever way you think is best. If `T.untyped` is described thusly then I am fine with using it

**Review by realityforge — APPROVED** [review]

**Pull request merged**